File: prettier-plugin-csharpier/src/Printer/Types/ParameterList.ts
import { printCommaList } from "../Helpers";
import { PrintMethod } from "../PrintMethod";
import { SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { ParameterNode } from "./Parameter";

export interface ParameterListNode extends SyntaxTreeNode<"ParameterList"> {
    parameters: ParameterNode[];
}

export const print: PrintMethod<ParameterListNode> = (path, options, print) => {
    const node = path.getValue();
    if (node.parameters.length === 0) {
        return "()";
    }

    return group(concat(["(", indent(concat([softline, printCommaList(path.map(print, "parameters"))])), ")"]));
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/ParameterList.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class ParameterList
{
    public static Doc Print(ParameterListSyntax node, PrintingContext context)
    {
        return Print(node, node.OpenParenToken, node.CloseParenToken, context);
    }

    public static Doc Print(
        BaseParameterListSyntax node,
        SyntaxToken openToken,
        SyntaxToken closeToken,
        PrintingContext context
    )
    {
        return Doc.Group(
            Token.Print(openToken, context),
            node.Parameters.Count > 0
                ? Doc.Concat(
                    Doc.Indent(
                        Doc.SoftLine,
                        SeparatedSyntaxList.Print(
                            node.Parameters,
                            Parameter.Print,
                            Doc.Line,
                            context
                        )
                    ),
                    Doc.SoftLine
                )
                : Doc.Null,
            Token.Print(closeToken, context)
        );
    }
}
